chicken-scheme: add package

jpellegrini opened the pull request:
Signed-off-by: Jeronimo Pellegrini <j_p@aleph0.info>

Compile tested: Run tested: on AR7xxx (TP-Link Archer C7 v.4)

Description:
This is an efficient Scheme interpreter, which comes with several
modules for networking, filesystem access, and other useful tasks.
It can be used as scripting language for automating tasks, by users
who prefer dynamic functional languages over imperative or
object-oriented ones.

Maintainer: me / jpellegrini 
Compile tested: compiles on x86_64, target mips (AR7xxx), OpenWRT master and OpenWRT 18.06.2
Run tested: on AR7xxx (TP-Link Archer C7 v.4).  Used the REPL, loaded libraries and used them.

Description:
This is a Scheme interpreter -- a very efficient one, which comes with several libraries for networking, filesystem access, parsing, and other tasks. It could be used as a scripting language by users who prefer dynamic functional languages over imperative or object-oriented ones.
Among all Scheme implementations available today, Chicken is one of the most popular, and has a reasonably large development team (it's not a one-man show). For more information, please see https://call-cc.org.

The .ipk file is 2.8Mb large, wnd would make sense to those using extroot.

yousong commented on lang/chicken-scheme/Makefile:
```diff
@@ -0,0 +1,53 @@
+endef
+
+define Build/Compile
+	$(MAKE) -C $(PKG_BUILD_DIR) PLATFORM=linux C_COMPILER=$(TARGET_CC) LINKER=$(TARGET_CC) $(TARGET_CONFIGURE_OPTS) PREFIX=/usr
```
Does it also accept `TARGET_CFLAGS` and other flags to allow for customizations from the base build system.  For example, optimize for size `-Os` , or other hardening flags in include/hardening.mk.

karlp replied:
You can also just provide MAKE_FLAGS="blah blah" instead of completely overwriting the compile rule, which will play better with parallel jobs.  You can look in the repo for other users of this.

karlp commented on lang/chicken-scheme/Makefile:
```diff
@@ -0,0 +1,53 @@
+PKG_BUILD_DIR:=$(BUILD_DIR)/chicken-$(PKG_VERSION)
+PKG_SOURCE:=chicken-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://code.call-cc.org/releases/5.0.0/
+PKG_MD5SUM:=de22ded475766fceaec3f9db2238ad76
```
Should be PGK_HASH and a sha256 sum these days.

karlp commented on lang/chicken-scheme/Makefile:
```diff
@@ -0,0 +1,53 @@
+define Package/chicken-scheme/install
+	$(MAKE) -C $(PKG_BUILD_DIR) DESTDIR=$(1) PLATFORM=linux PREFIX=/usr C_COMPILER=$(TARGET_CC) LINKER=$(TARGET_CC) install
+	$(RM) -rf $(1)/usr/share/man $(1)/usr/share/chicken/doc
+	$(RM) -rf $(1)/usr/share/include $(1)/usr/bin/csc $(1)/usr/bin/chicken-install $(1)/usr/bin/chicken-uninstall $(1)/usr/bin/chicken-profile $(1)/usr/bin/chicken
```
do they really not have make targets that skip building the docs?

jpellegrini commented:
Thanks for the comments! I have addressed those in the second commit (`916d07...`).

yousong commented on lang/chicken-scheme/Makefile:
```diff
@@ -0,0 +1,66 @@
+		$(1)/usr/bin/feathers \
+		$(1)/usr/bin/chicken-status \
+		$(1)/usr/bin/chicken-do \
+		$(1)/usr/lib/libchicken.a 
```
Maybe we should just use `$(CP)` to select what to include.  That way it will be more clear to the reader what's actually included in the package, not the other way around.  `$(CP)` can also error out more promptly should something be missing in future generations of chickens.  And if we decide to also package these removed parts in future versions, the code will be more consistent.

yousong commented:
@jpellegrini , thanks for the work.  Please squash these commits into a single one when you are ready.

jpellegrini commented:
Okay, squashed and pushed.

- switched from mv to install+cp, as suggested
- gave the package a more descriptive name.

Thank you for you review!

yousong commented:
Thanks @jpellegrini  !
